### PR TITLE
Allow to disable on pre-prod environments

### PR DESCRIPTION
**Pull request by jules0x**
Restoring PROD data to UAT can result in the CMS user receiving emails from both environments. Allow to disable for pre-prod environments via config.

### CODE REVIEW

**Comment by michalkleiner** [comment]
This could possibly be turned into a generic `send_emails` config defaulting to true where you then use config system to disable it for non-Live environments. That way people would be able to disable emails completely, if they wanted that for whatever reason, not just in preprod envs.

Note that config properties use snake_case.

**Comment by jules0x** [comment]
@michalkleiner that makes sense, I'll update it now. 

In my current use case, I now need to provide a control for the user within the CMS to toggle it on and off. 
I'm considering a SiteConfig extension adding a boolean class, which is presented to the user only if the **send_emails** is set to _true_, and an additional static **cms_email_toggle** (or similar) is set to _true_
It feels like that may be too opinionated a change to bring into the module. Do you agree?

File: src/Tasks/ContentReviewEmails.php
Comment (michalkleiner) [inline src/Tasks/ContentReviewEmails.php:33]: This message is not accurate, the task hasn't been disabled, the review email sending has been. I'm not sure we should output a message here in this case, might be better to log an info message. Do you know how other tasks handle output of this type?

**Comment by michalkleiner** [comment]
> In my current use case, I now need to provide a control for the user within the CMS to toggle it on and off. I'm considering a SiteConfig extension adding a boolean class, which is presented to the user only if the **send_emails** is set to _true_, and an additional static **cms_email_toggle** (or similar) is set to _true_ It feels like that may be too opinionated a change to bring into the module. Do you agree?

Yeah, that would be too opinionated. In that case, you'd disable the default task and then subclass the task and reimplement the run method in any way you like, reflecting your site settings etc.

File: src/Tasks/ContentReviewEmails.php
Comment (jules0x) [reply src/Tasks/ContentReviewEmails.php:33]: Good call. No message is required here. 
Perhaps a LiteralField on the content review tab within the page settings could indicate the email configuration state.
image

Comment (michalkleiner) [reply src/Tasks/ContentReviewEmails.php:33]: Let's just ditch this echo to get the PR in and then we can deal with the UI message in a separate enhancement.

File: docs/en/userguide/index.md
Comment (ChloeSartorelli) [inline docs/en/userguide/index.md:37]:
```suggestion
---
Except:
  environment: live
---
SilverStripe\ContentReview\Tasks\ContentReviewEmails:
  send_emails: false
```

File: src/Tasks/ContentReviewEmails.php
Comment (ChloeSartorelli) [reply src/Tasks/ContentReviewEmails.php:33]: The sending of the email is all this task does, right? So disabling the sending of content review emails is effectively the same as disabling this task.

It seems to me that what we really want is a way to disable any `BuildTask` - which would be an enhancement to framework.

**Review by GuySartorelli — CHANGES_REQUESTED** [review]

File: src/Tasks/ContentReviewEmails.php
Comment (michalkleiner) [reply src/Tasks/ContentReviewEmails.php:33]: Yep, we can look at doing a config enhancement to disable individual tasks, I think there was something happening already elsewhere, but that shouldn't block the PR here.

Comment (ChloeSartorelli) [reply src/Tasks/ContentReviewEmails.php:33]: That did actually get merged: https://github.com/silverstripe/silverstripe-framework/pull/10801
I'd argue that if the intention is just to not send email, and if this task effectively does nothing other than send email, then the merged feature to disable tasks makes this PR redundant.

That said if I've missed something here I'm happy for this PR to continue forward.

Comment (michalkleiner) [reply src/Tasks/ContentReviewEmails.php:33]: Right. That will only become available in 5.1, not available for CMS 4. Or can we back-port that? Even then it's probably another 4.x minor so not within easy reach compared to having this immediately available here. Or are we constraint here not doing a module minor while waiting for CMS minor to keep in sync?

I would personally prefer to get this in and then look at updating tasks across kitchen sink to that mechanism if there were more of them that could get refactored from custom mechanism to this new approach.

Comment (michalkleiner) [reply src/Tasks/ContentReviewEmails.php:33]: Just opened https://github.com/silverstripe/silverstripe-contentreview/pull/201 which can ship in patch here and then we can close this PR completely and wait for the new task disablement mechanism in 5.1. Backporting to 4.x question probably still valid there.

Comment (ChloeSartorelli) [reply src/Tasks/ContentReviewEmails.php:33]: No new features can be in CMS 4 branches for supported modules, since we're not releasing any new minors for CMS 4

Comment (michalkleiner) [reply src/Tasks/ContentReviewEmails.php:33]: Ok, was anticipating that. So the patch update here in the other PR would solve it for now.

Comment (ChloeSartorelli) [reply src/Tasks/ContentReviewEmails.php:33]: Yup

**Comment by michalkleiner** [comment]
Thanks for the contribution, @jules0x.

After further discussion within the core team this will be achievable by using the core mechanism introduced in https://github.com/silverstripe/silverstripe-framework/pull/10801 that will be available from CMS 5.1

Until then, using [4.7.1](https://github.com/silverstripe/silverstripe-contentreview/releases/tag/4.7.1), `Injector` can be used to replace the task with the custom config and custom logic in the `run` method of the task.

With that, I'll close the PR for now.

**Pull request closed without merging**
